feat: 인기 태그 추천 구현

HeungY opened the pull request:
# 인기 태그 추천 구현

### 개요
> 변경된 부분을 짧게 요약해주세요.

HotTagManager
- generate 도메인 내부에 인기있는 태그 목록 반환하는 /tag 엔드포인트 구현
- Scheduled 어노테이션 통해 매 시간 정각마다 DB의 생성된 태그 가져와서 인기 순으로 정렬, 최대 10개의 태그까지 HotTag로 분류

ShedLock 라이브러리
- 분산 환경에서 스케줄링이 중복으로 실행되는 것을 막기 위해 ShedLock 라이브러리 도입했습니다.

### 리뷰어가 꼭 봐줬으면 하는 부분
> 코드의 특정 부분을 지정해주셔도 좋고, 커밋을 지정해주셔도 좋고,   
> 로직적으로 제시해주셔도 좋습니다.

HotTagManager
매 시간마다 생성하는데 GC가 참조되지 않는 데이터 정리해줄거라 생각하므로 메모리는 크게 걱정되지는 않지만 혹시 문제가 생긴다면 변경하겠습니다.

### Jira ISSUE Keys
> 관련된 Jira 백로그 키를 나열해주세요. 여러개 있다면 여러개 작성해주세요
- PQ-80

Commit:
feat: 인기 태그 추천 구현

PQ-80

HotTagManager
- generate 도메인 내부에 인기있는 태그 목록 반환하는 /tag 엔드포인트 구현
- Scheduled 어노테이션 통해 매 시간 정각마다 DB의 생성된 태그 가져와서 인기 순으로 정렬, 최대 10개의 태그까지 HotTag로 분류

매 시간마다 생성하는데 GC가 참조되지 않는 데이터 정리해줄거라 생각하므로 메모리는 크게 걱정되지 않습니다.

carter468 committed:
feat: 나의 4Q 리스트 응답에 태그명, 배경이미지 좋아요 개수 정보 추가

PQ-27 - PQ-53 - PQ-115

나의 4Q 리스트 응답에 태그명, 배경이미지 좋아요 개수 정보 추가
- 태그명과 배경이미지 매핑해서 쿼리수 감소
- MyPageController 일반예외핸들러 삭제
- 엔드포인트 myPQ -> mypq 소문자로 변경

ZZ3n committed:
hotfix: tag engName

Commit:
fix: 이메일 중복 검사

PQ-131

EmailController
- /users/email/verification
-> 이메일 중복시, CONFLICT 반환.

carter468 committed:
fix: categoryName으로 필터링할때 오류나는 버그 수정

carter468 committed:
fix: categoryName으로 필터링할때 오류나는 버그 수정

Commit:
fix: OAuth 로그인 후 이메일 로그인 방지

PQ-131

UserService
- /users/login 엔드포인트
- signUpOrLogin 객체
-> DB의 oauthProvider와 입력받은 oauthProvider가 다를시, 에러 반환

Commit:
feat: 비밀번호 재설정 구현

PQ-130

EmailController
- /users/email/update
-> 이메일 인증 후 DB에 해당하는 이메일이 있을시, 정상 응답

UserController
- /users/reset
-> 비밀번호 재설정.

/users/email/update로 이메일 확인 하고, /users/reset 으로 이메일과 새로운 비밀번호 입력하면 비밀번호 재설정 완료됩니다.

Commit:
feat: 비밀번호 재설정 구현

PQ-130

EmailController
- /users/email/update
-> 이메일 인증 후 DB에 해당하는 이메일이 있을시, 정상 응답

UserController
- /users/reset
-> 비밀번호 재설정.

/users/email/update로 이메일 확인 하고, /users/reset 으로 이메일과 새로운 비밀번호 입력하면 비밀번호 재설정 완료됩니다.

Commit:
fix: void 타입 제거

PQ-130

UserService
- resetPassword() 객체 반환타입 void
-> DTO로 변경하여 id내려줌.

Commit:
feat: 회원정보 수정 구현

PQ-130

UserController
- /users/modify 엔드포인트로 nickname PATCH 요청
UserService
- modify() 객체
- 현재 유저의 정보를 받아와서 이메일을 통해 DB 정보 반환.
- nickname 수정 후 DB에 덮어씀

Commit:
feat: 회원정보 수정 구현

PQ-130

UserController
- /users/modify 엔드포인트로 nickname PATCH 요청
UserService
- modify() 객체
- 현재 유저의 정보를 받아와서 이메일을 통해 DB 정보 반환.
- nickname 수정 후 DB에 덮어씀

Commit:
fix: 회원 정보 수정 오류 수정

PQ-129

ModifyRequestDTO
- 기본 생성자 추가해서 기존 초기화 안되던 오류 수정했습니다.

Commit:
fix: 회원 정보 수정 오류 수정

PQ-129

ModifyRequestDTO
- 기본 생성자 추가해서 기존 초기화 안되던 오류 수정했습니다.

carter468 committed:
feat: 마이페이지 내가 좋아요한 4Q 응답 구현

PQ-27 - PQ-38

마이페이지 내가 좋아요한 4Q 응답 구현

ZZ3n commented on src/main/java/org/chunsik/pq/generate/manager/HotTagManager.java:
```diff
@@ -0,0 +1,53 @@
+        updateTags();
+    }
+
+    @Scheduled(cron = "0 0 * * * *")
```
저희 서버가 여러개일 경우를 대비해야 합니다...
지금은 1개이기는한데, 나중에 늘어날 계획이라서 그렇습니다.

---
## ChatGPT
@Scheduled는 Spring 프레임워크에서 일정한 주기마다 특정 메서드를 실행하기 위해 사용됩니다. 그러나 인스턴스가 여러 개 있을 경우, 분산 환경에서 동일한 작업이 여러 인스턴스에서 중복으로 실행될 수 있는 문제가 발생할 수 있습니다. 이러한 문제를 해결하려면 아래와 같은 방법들을 고려할 수 있습니다.

1. 싱글 인스턴스에서만 실행하도록 설정

여러 인스턴스 중 한 곳에서만 @Scheduled가 실행되도록 설정하는 방법입니다. 이를 위해 대표적으로 사용하는 방법은 다음과 같습니다:

	•	Database Lock 사용: 하나의 작업이 이미 실행 중임을 데이터베이스에 기록하여 중복 실행을 방지합니다.
	•	Redis와 같은 분산 락 사용: 분산 환경에서 락을 사용하여 중복 실행을 방지할 수 있습니다. Redis의 SETNX 명령을 사용하여 락을 걸고, 락이 존재하지 않을 때만 작업을 실행하도록 설정할 수 있습니다.

2. Spring Cloud와 Leader Election 사용

Spring Cloud와 같은 도구를 사용하면 여러 인스턴스 중 리더 인스턴스를 선출하여 리더 인스턴스에서만 @Scheduled 작업을 실행할 수 있습니다. Spring Cloud에서의 리더 선출을 위해 Spring Cloud Consul이나 Spring Cloud Zookeeper와 같은 기술을 사용할 수 있습니다.

3. 분산 작업 스케줄링 도구 사용

@Scheduled는 단순한 스케줄링이 필요한 경우에 적합하지만, 분산 환경에서 좀 더 복잡한 스케줄링 관리가 필요할 경우에는 Quartz나 Spring Batch 같은 분산 작업 관리 도구를 사용하는 것이 더 적합할 수 있습니다.

결론적으로, 인스턴스가 여러 개 있을 때는 @Scheduled만으로는 중복 실행 문제를 해결할 수 없으므로 분산 락, 리더 선출, 또는 다른 스케줄링 도구와 같은 보완책을 사용하는 것이 좋습니다.

ZZ3n replied:
해결방법은 GPT한테 함 물어봤어요

ZZ3n reviewed (changes requested):
분산 인스턴스 환경에서 Schduled 를 사용하면 문제가 있을 수 있습니다.

Commit:
fix: 비밀번호 변경 Setter 제거

PQ-130

UserService
- 기존 Setter 사용하여 비밀번호 업데이트
-> User 엔티티 내부 resetPassword 객체 생성하여 업데이트로 변경

kimdaegeon0918 commented on src/main/java/org/chunsik/pq/generate/manager/HotTagManager.java:
```diff
@@ -0,0 +1,53 @@
+@RequiredArgsConstructor
+public class HotTagManager {
+    private final TagRepository tagRepository;
+    private final TagBackgroundImageRepository tagBackgroundImageRepositoryRepository;
```
변수명에 Repository가 두번들어갔습니다.

carter468 reviewed (commented):
코드 잘 보았습니다 👍

Commit:
fix: 닉네임 변경 Setter 제거

PQ-130

UserService
- 기존 Setter 사용하여 닉네임 업데이트
-> User 엔티티 내부 modifyNickname 객체 생성하여 업데이트로 변경

Commit:
fix: 비밀번호 재설정 이메일 인증 로직 수정

PQ-130

EmailController
- /update 엔드포인트를 /verification 엔드포인트에 결합

겹치는 부분이 많아 로직 합치고 수정했습니다. 추가로 이메일 중복, 이메일 없을경우 예외처리 추가해줬습니다.

Commit:
fix: null 체크 중복 제거

PQ-130

ModifyRequestDTO
- @NotBlank가 NotNull의 역할도 하기때문에 @NotNull 제거

ZZ3n committed:
feat: add login expiration time

PQ-132

- login response 수정
- login service DTO 수정

carter468 committed:
chore: NotBlank 두번 사용한것 삭제

Commit:
fix: EmailController 엔드포인트 분리

PQ-130
EmailController
- 하나의 엔드포인트에서 두개의 로직 수행중이던 코드 변경
-> /users/email/verification
-> /users/email/reset
두개의 엔드포인트로 분리
ModifyRequestDTO
- A-Z 추가

ZZ3n committed:
fix: 비밀번호 초기화 인증 관련 convention 수정

PQ-130

- AESUtil Exception 좀 더 좁은 범위로 수정
- 쓰이지 않는 Exception 삭제
- early return 으로 변경

Commit:
chore: resolve conflict

Commit:
chore: resolve conflict

Commit:
chore: resolve conflict

HeungY committed:
Merge branch 'dev' into feature/hottag

Commit:
chore: TagBackgroundImageRepository 이름 수정

Commit:
feat: 분산 환경 스케줄링 구현

PQ-80

ShedLock 라이브러리 사용하여 분산 환경에서 하나의 스케줄러만 동작하도록 구현하였습니다.

Commit:
chore: 코드 정렬

Commit:
fix: resolve conflict

carter468 reviewed (approved):
LGTM 👍

ZZ3n committed:
Merge branch 'dev' into feature/hottag
